src/leveled_tictac.erl

```diff
@@ -502,48 +516,71 @@ to_level1_map_loop(<<>>, L1MapAcc, _Idx) ->
 to_level1_map_loop(<<Slice:64/binary, Rest/binary>>, L1MapAcc, Idx) ->
     to_level1_map_loop(Rest, maps:put(Idx, Slice, L1MapAcc), Idx + 1).
 
-
 -spec from_level1_map(level1_map()) -> binary().
-from_level1_map(L1Map) ->
+from_level1_map(Level1M) when is_map(Level1M) ->
     lists:foldl(
         fun(I, Acc) ->
-            <<Acc/binary, (maps:get(I, L1Map))/binary>>
+            <<Acc/binary, (maps:get(I, Level1M))/binary>>
         end,
         <<>>,
-        lists:seq(0, maps:size(L1Map) - 1)
-    ).
+        lists:seq(0, maps:size(Level1M) - 1)
+    );
+from_level1_map(Level1B) when is_binary(Level1B) ->
+    Level1B.
 
 -spec extract_segment(
     integer(), tictactree()) -> 
         {integer(), integer(), tree_extract(), tree_extract()}.
 %% @doc
 %% Extract the Level 1 and Level 2 slices from a tree to prepare an update
 extract_segment(Segment, TicTacTree) ->
-    Level2Pos =
-        Segment band (?L2_CHUNKSIZE - 1),
     Level1Pos =
         (Segment bsr ?L2_BITSIZE)
             band (TicTacTree#tictactree.width - 1),
-    Level1Slice = Level1Pos div 16,
     
+    Level2Pos =
+        Segment band (?L2_CHUNKSIZE - 1),
     Level2BytePos = ?HASH_SIZE * Level2Pos,
-    Level1BytePos = ?HASH_SIZE * (Level1Pos rem 16),
-    
     Level2 = get_level2(TicTacTree, Level1Pos),
     
     HashIntLength = ?HASH_SIZE * 8,
     <<PreL2:Level2BytePos/binary,
         SegLeaf2:HashIntLength/integer,
         PostL2/binary>> = Level2,
-    <<PreL1:Level1BytePos/binary,
-        SegLeaf1:HashIntLength/integer,
-        PostL1/binary>> = maps:get(Level1Slice, TicTacTree#tictactree.level1),
+    {PreL1, SegLeaf1, PostL1, Level1BytePos} =
+        extract_level1_slice(TicTacTree#tictactree.level1, Level1Pos),
     
     {SegLeaf1, 
         SegLeaf2, 
         {PreL1, Level1BytePos, Level1Pos, HashIntLength, PostL1},
         {PreL2, Level2BytePos, Level2Pos, HashIntLength, PostL2}}.
 
+-spec extract_level1_slice(
+    level1_map(), non_neg_integer()) ->
+        {binary(), non_neg_integer(), binary(), non_neg_integer()}.
+extract_level1_slice(Level1M, Level1Pos) when is_map(Level1M) ->
+    Level1Slice = Level1Pos div 16,
+    HashIntLength = ?HASH_SIZE * 8,
+    Level1BytePos = ?HASH_SIZE * (Level1Pos rem 16),
+    <<PreL1:Level1BytePos/binary,
+        SegLeaf1:HashIntLength/integer,
+        PostL1/binary>> = maps:get(Level1Slice, Level1M),
+    {PreL1, SegLeaf1, PostL1, Level1BytePos};
+extract_level1_slice(Level1B, Level1Pos) when is_binary(Level1B) ->
+    HashIntLength = ?HASH_SIZE * 8,
+    Level1BytePos = ?HASH_SIZE * Level1Pos,
```

Comment:
Why the difference with the original code where you do:
`Level1BytePos = ?HASH_SIZE * (Level1Pos rem 16)`

Ah, I see... the original code has this, then you changed it to `rem 16` for the newer version, but the old version sticks with the value you have here.